src/_guides/language/effective-dart/design.md

```diff
@@ -1005,90 +1091,262 @@ var names = people.map((Person person) {
+static type for you. That word "most" is the alarming part. Sometimes, inference
+fails and silently uses `dynamic` instead. If you didn't realize that inference
+failed, you now have code that *looks* statically safe but actually has no
+type checking whatsoever.
 
-In most places in Dart, a type annotation can be omitted, in which case the type
-will automatically be `dynamic`. Thus, omitting the type annotation entirely is
-semantically equivalent but more terse.
+Sometimes, `dynamic` is what you want. Dart does support a dynamic type for good
+reasons. But a casual reader of your code who sees an annotation is missing has
+no way of knowing if you intended it to be `dynamic`, expected inference to fill
+in some other type, or you simply forgot to write the annotation.
+
+Using `dynamic` explicitly when that is the type you want makes that intention
+clear.
 
 {:.good-style}
-<?code-excerpt "misc/lib/effective_dart/design_good.dart (avoid-dynamic)"?>
+<?code-excerpt "misc/lib/effective_dart/design_good.dart (prefer-dynamic)"?>
 {% prettify dart %}
-lookUpOrDefault(String name, Map map, defaultValue) {
+T lookUp<T>(String name, Map<String, T> map, T defaultValue) {
   var value = map[name];
   if (value != null) return value;
   return defaultValue;
 }
 {% endprettify %}
 
 {:.bad-style}
-<?code-excerpt "misc/lib/effective_dart/design_bad.dart (avoid-dynamic)"?>
+<?code-excerpt "misc/lib/effective_dart/design_bad.dart (prefer-dynamic)"?>
 {% prettify dart %}
-dynamic lookUpOrDefault(String name, Map map, dynamic defaultValue) {
+lookUp(String name, Map map, defaultValue) {
   var value = map[name];
   if (value != null) return value;
   return defaultValue;
 }
 {% endprettify %}
 
 
-### AVOID annotating with `Function`.
+<aside class="alert alert-info" markdown="1">
 
-The `Function` type is barely more precise than using no annotation at all. If
-you're bothering to annotate, it's better to use a precise function type that
-describes the signature and return type of the function.
+Before Dart 2, this guideline stated the exact opposite: *don't* annotate with
+`dynamic` when it is implicit. With the new stronger type system and type
+inference, users now expect Dart to behave like an inferred statically-typed
+language. With that mental model, it is an unpleasant surprise to discover that
+a region of code has silently lost all of the safety and performance of static
+types.
 
-If you are annotating a field, this does mean you have to create a typedef, but
-that's usually worth doing.
+</aside>
+
+
+### PREFER signatures in function type annotations.
+
+The identifier `Function` by itself without any return type or parameter
+signature refers to the [special `Function` type][fn class]. This type is only
+marginally more useful than using `dynamic`. If you're going to annotate, prefer
+a full function type that includes the parameters and return type of the
+function.
+
+[fn class]: https://api.dartlang.org/stable/dart-core/Function-class.html
```

Comment:
This needs to be:

    [Function]: {{site.dart_api}}/{{site.data.pkg-vers.SDK.channel}}/dart-core/Function-class.html

Reply:
Done.